_posts/2021/2021-09-28-cpp-lessons-learned.md

```diff
@@ -0,0 +1,147 @@
+---
+title: Libpmemobj-cpp - lessons learned
+author: lukaszstolarczuk
+layout: post
+identifier: cpp-lessons-learned
+---
+
+### Introduction
+
+We've been working on C++ bindings for *libpmemobj* since around 2016 - see our [very first
+tutorial for libpmemobj-cpp][cpp_tutorial_blog]. We've come a long way since then. A lot has
+changed - we've gained more experience and knowledge, added new features, fixed quite a few
+bugs, and created at least half a dozen new containers. It's fair to state this product is now
+far more mature and well-developed. Over time, we've learned at least several lessons about
+designing and overcoming issues in C++ applications for persistent memory. We want to share
+them with you to learn from **our** mistakes and not waste any time.
+
+### Design problems
+
+Designing efficient concurrent algorithms for persistent memory is a challenging task.
+Actually, this is one of the reasons why we introduced libpmemobj-cpp in the first place.
+We wanted to do the hard work once, test it as much as possible, and give people reliable,
+efficient, ready-to-use solutions. We describe below only a few of our problems, which
+we believe might benefit for a broader audience.
+
+#### Optimize number of allocations!
+
+Lots of allocations on pmem may ruin performance. It seems like an obvious statement, but
+it's easy to overlook this issue. One badly placed allocation (e.g., in critical
+code path), and we end up with low performance. Unfortunately, inserting a new node in
+`pmem::obj::concurrent_hash_map` is associated with multiple allocations.
+When we were thinking about the next container to implement, we chose radix tree structure.
+It is a great fit for pmem, because it requires a minimal number of allocations, no tree
+re-balancing, and it is sorted but requires no key comparisons.
+We realized `pmem::obj::experimental::radix_tree` can be optimized even further.
+To store keys or values of variable length, `string` (`pmem::obj::string` to be precise)
+is a natural choice. Each such string actually equals two allocations.
```

Comment:
That's not entirely true. Each string (potentially, because there is also small string optimization) equals one extra allocation (not two)